deploy/tests/test_generate_production_ini.py

```diff
@@ -0,0 +1,112 @@
+import os
+import pytest
+
+from contextlib import contextmanager
+from io import StringIO
+from unittest import mock
+
+from .. import generate_production_ini
+from ..generate_production_ini import (
+    TEMPLATE_DIR,
+    build_ini_file_from_template,
+    build_ini_stream_from_template,
+    environment_template_filename,
+    template_environment_names,
+)
+
+@contextmanager
+def override_environ(**overrides):
+    to_delete = []
+    to_restore = {}
+    env = os.environ
+    try:
+        for k, v in overrides.items():
+            if k in env:
+                to_restore[k] = env[k]
+            else:
+                to_delete.append(k)
+            env[k] = v
+        yield
+    finally:
+        for k in to_delete:
+            del os.environ[k]
+        for k, v in to_restore.items():
+            os.environ[k] = v
+
+def test_environment_template_filename():
+
+    with pytest.raises(ValueError):
+        environment_template_filename('foo')
+
+    actual = os.path.abspath(environment_template_filename('webdev'))
+
+    assert actual.endswith("/ini_files/webdev.ini")
+    assert os.path.exists(actual)
+
+    assert environment_template_filename('webdev') == environment_template_filename('fourfront-webdev')
+
+
+def test_template_environment_names():
+
+    names = template_environment_names()
+
+    required_names = ['hotseat', 'mastertest', 'webdev', 'webprod', 'webprod2']
+
+    for required_name in required_names:
+        assert required_name in names
+
+
+def test_build_ini_file_from_template():
+    # NOTE: This implicitly also tests build_ini_file_from_stream.
+
+    lines = []
+    some_template_file_name = "mydir/whatever"
+    some_ini_file_name = "mydir/production.ini"
+    output_string_stream = StringIO()
+    env_vars = dict(RDS_DB_NAME='snow_white', RDS_USERNAME='user', RDS_PASSWORD='my-secret',
+                    RDS_HOSTNAME='unittest', RDS_PORT="6543")
+
+    with override_environ(**env_vars):
+
+        for env_var in env_vars:
+            assert env_var in os.environ and os.environ[env_var] == env_vars[env_var], (
+                    "os.environ[%r] did not get added correctly" % env_var
+            )
+
+        class MockFileStream:
+            def __init__(self, filename, mode):
+                pass
+            def __enter__(self):
+                return output_string_stream
+            def __exit__(self, type, value, traceback):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Redefining built-in 'type'](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5189083)